FreeRTOS/FreeRTOS-Plus-TCP
# IPv4/single: connect() should return immediately after a protocol error and other things

Description
-----------
A TCP connection can be established in either two ways: accept() and connect().

When using connect(), the socket is created and owned by the application.

When using accept(), the new socket is created and temporarily owned by the IP-task, and in the comments it is called an "orphaned socket". Once the connection is established, the function accept() will pass the ownership to the application.

The stack has a mechanism that checks sockets for inactivity: `ipconfigTCP_HANG_PROTECTION`. After `ipconfigTCP_HANG_PROTECTION_TIME` seconds of inactivity an orphaned socket will be closed.

There were a few reasons to create this PR:

1) When a protocol error occurs during a `FreeRTOS_connect()`, the API should stop waiting and return `-pdFREERTOS_ERRNO_ENOTCONN` to the application. At this moment it will only return when the timeout has been reached.

2) When a protocol error occurs during a passive connection, the socket will hang until the "hang protection" will close the socket. When a fatal protocol error occurs the client socket should be closed immediately.

3) The re-use option `FREERTOS_SO_REUSE_LISTEN_SOCKET`: when the listening socket is configured to turn itself into a "client socket", it may not be closed after a protocol error or after "hang protection". The reusable socket will be put into the listening state again and it will be able to receive a new connection.


Many thanks to [Am Paschal](https://forums.freertos.org/u/AmPaschal), who wrote the post "FreeRTOS connect and accept functions (in FreeRTOS+TCP) do not return even after the TCP connection closes" on the [FreeRTOS forum](https://forums.freertos.org/t/freertos-connect-and-accept-functions-in-freertos-tcp-do-not-return-even-after-the-tcp-connection-closes).

Also thanks to [Alex "oberondoro"](https://forums.freertos.org/u/oberondoro) who wrote the post "Sockets end in state eCLOSE_WAIT" on the [FreeRTOS forum](https://forums.freertos.org/t/sockets-end-in-state-eclose-wait).

And thanks to [Hubert Sack](https://forums.freertos.org/u/hs4FreeRtos) who explained the problem which occurs when a reusable socket has a protocol error or hang in the SYN phase.

Test Steps
-----------

Testing this PR was a bit complicated because of the many possibilities:

1) Test an outgoing connection using connect()
2) Test an incoming connection with a normal listening socket
3) Test an incoming connection with a reusable socket

X. Induce a protocol error during the SYN phase
Y. Induce a time during the SYN phase and get a hang protection.

I did 6 tests in total: 1X, 2X, 3X, and 1Y, 2Y, and 3Y.

When the listening socket is reusable, it should return to its listening state and be able to receive a new connection.
Normal client sockets should be closed when a protocol error or a timeout occurs so that the resources become available again.
`FreeRTOS_connect()` should return immediately when a protocol error occurs, or when the socket stays inactive for a long time.


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
